cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/KafkaCruiseControlConfig.java

```diff
@@ -895,6 +895,13 @@
   public static final String WEBSERVER_SESSION_PATH = "webserver.session.path";
   private static final String WEBSERVER_SESSION_PATH_DOC = "Default Session Path (for cookies)";
 
+  /**
+   * <code>security.provider</code>
+   */
+  public static final String SECURITY_PROVIDER = "security.provider";
+  private static final String SECURITY_PROVIDER_DOC = "SecurityProvider implementation for defining authentication " +
+          "and authorization rules";
+
```

Comment:
Sure, I'll add HTTP in the following days.

To the other part: my code doesn't contain any kind of implementation, just the test have a very simplistic one. Do you think we would need to add a HTTP Basic implementation? Perhaps the biggest question in this case is what should be the credential source or user database?

Reply:
Thanks for the updates on `HTTPS` side!

> To the other part: my code doesn't contain any kind of implementation, just the test have a very simplistic one. Do you think we would need to add a HTTP Basic implementation?

Yes, I agree that we should have a HTTP Basic implementation to process incoming requests (if `enable.basic.authentication=true`). In the initial iteration, we can have a simple implementation -- e.g. just `ADMIN`, `USER`, and `VIEWER` roles with access to (1) all endpoints, (2) `GET` endpoints, and (3) `sync` `GET` endpoints.

> Perhaps the biggest question in this case is what should be the credential source or user database?

I propose starting simple for this part. Overall, maybe we can assume that
1. `WEBSERVER_SSL_KEYSTORE_LOCATION` always contains the certificate corresponding to `WEBSERVER_SSL_KEY_PASSWORD`, `WEBSERVER_SSL_KEYSTORE_TYPE`, and `WEBSERVER_SSL_KEYSTORE_PASSWORD`.
2. We can have configs corresponding to valid `user name` - `password` pairs (i.e. in `.properties` file). The deployment system configures this file such that user access to it is prevented. Going forward, we can have a followup patch to make the retrieval of user names and passwords pluggable; hence, a user database can be used to retrieve credentials.

-- Also adding a config to enable/disable basic auth would also let us add a sanity check in `KafkaCruiseControlConfig` to ensure that all relevant configs are set in case basic auth is required -- i.e.:
* `WEBSERVER_SSL_KEYSTORE_LOCATION`,
* `WEBSERVER_SSL_KEYSTORE_PASSWORD`
* `WEBSERVER_SSL_KEYSTORE_TYPE`
* `WEBSERVER_SSL_KEY_PASSWORD`

Reply:
Thanks for the detailed answer, I have two questions regarding the following:
> e.g. just ADMIN, USER, and VIEWER roles with access to (1) all endpoints, (2) GET endpoints, and (3) sync GET endpoints.
So in this case (3) would be essentially those requests which don't have a cookie or session ID? 
Also as I saw there are four endpoint types in CC which are: `KAFKA_MONITOR`, `CRUISE_CONTROL_MONITOR`, `KAFKA_ADMIN`, `CRUISE_CONTROL_ADMIN`. Looking at these the `*_MONITOR` endpoints are mostly the `GET` ones and the `*_ADMIN` are mostly the `POST` requests. There are two exceptions though, `bootstrap` and `train` which are both `GET` but of `CRUISE_CONTROL_ADMIN` type. Would it make sense to treat them as they belong to the `ADMIN` role and won't be able to be accessed by the other two?

Reply:
> So in this case (3) would be essentially those requests which don't have a cookie or session ID?

My initial access control idea was to have a separation between `resource-intensive` vs. `light` calls, where `sync GET endpoints` would be the light and safe calls. However, maybe roles based on `CruiseControlEndpointType` is a better idea.

`bootstrap` and `train` are `GET` due to legacy reasons (and are rarely used today) -- they should actually require `ADMIN` powers.

Reply:
An update here: as I see then `viewer` will have access only to `kafka_cluster_state`, `user_tasks` and `review_board` endpoints. The rest of the GET endpoints are async, thus they require `user` role. Then `bootstrap` and `train` although are GET endpoints will require `admin` with the rest of the requests. Also `admin` can access whatever `user` and `viewer` can.

Reply:
Yes, that sounds great to me!